cli/cycpp.py

```diff
@@ -1045,224 +1084,236 @@ def _val(self, t, val=None, name=None, uitype=None):
-        v = 'std::pair<{0}, {1} >({2}, {3})'
-        v = v.format(type_to_str(ftype), type_to_str(stype), first, second)
+        uitype = prepare_type(t, uitype)
+
+        fname, sname = 'first', 'second'
+        first = self._val(ftype, val=val[0], uitype=uitype[1], name=fname)
+        second = self._val(stype, val=val[1], uitype=uitype[2], name=sname)
+        v = '{0} {1};'.format(type_to_str(t), name)
+        v += '{'
+        v += '{0} {1} {2}.first = {3}; {2}.second = {4};'.format(
+                first, second, name, fname, sname)
+        v += '}'
         return v
 
     def _val_map(self, t, val=None, name=None, uitype=None):
-        val = val or {}
-        if not isinstance(val, Mapping):
-            val = dict(val)
-        ktype, vtype = t[1], t[2]
-        kuitype = None if uitype is None else uitype[1]
-        vuitype = None if uitype is None else uitype[2]
-        v = ''
+        uitype = prepare_type(t, uitype)
+        keyname = 'key'
+        valname = 'val'
+
+        v = '{0} {1};'.format(type_to_str(t), name)
+        v += '{'
         for k, x in val.items():
-            k = self._val(ktype, val=k, uitype=kuitype)
-            x = self._val(vtype, val=x, uitype=vuitype)
-            v += '{0}[{1}] = {2};\n'.format(name, k, x)
+            kval = self._val(t[1], val=k, uitype=uitype[1], name=keyname)
+            xval = self._val(t[2], val=x, uitype=uitype[2], name=valname)
+            v += '{{ {0} {1} {2}[{3}] = {4};}}'.format(kval, xval, name, keyname, valname)
+        v += '}'
+
         return v
 
-    def _query(self, tree, alias, t, d, uitype=None, idx=None):
+    def _query(self, tree, alias, t, uitype=None, idx=None):
         tstr = type_to_str(t)
         if tstr.endswith('>'):
             tstr += " "
         # Get keys
         kw = {'cycns': CYCNS, 'type': tstr, 'alias': alias, 'tree': tree}
-        if d is None:
-            kw['query'] = "Query"
-            kw['default'] = ""
-        else:
-            kw['query'] = "OptionalQuery"
-            default = self._val(t, val=d, uitype=uitype)
-            if ';' in default:
-                raise ValueError('can only query based on values that '
-                                 'are C++ expressions, got:\n\n' + default)
-            kw['default'] = ", " + default
         kw['index'] = '' if idx is None else ', {0}'.format(idx)
         # get template
         if uitype == 'nuclide':
-            template = ('pyne::nucname::id({cycns}::{query}<std::string>({tree}, '
-                        '"{alias}"{default}{index}))')
-            if d is not None:
-                kw['default'] = ', "{0}"'.format(d)
+            template = ('pyne::nucname::id({cycns}::Query<std::string>({tree}, '
+                        '"{alias}"{index}))')
         else:
-            template = '{cycns}::{query}<{type}>({tree}, "{alias}"{default}{index})'
+            template = '{cycns}::Query<{type}>({tree}, "{alias}"{index})'
```

Comment:
If the statevar is optional, the top level impl, now catches this in an if block.  It checks to see if the variable was defined in the schema before running this cpp code.  So This code only ever runs after we know that the optional value has *already* been defined in the input file.  So it was actually a redundant check before.  We don't need the default, because that is constructed and assigned by code that gets inserted into the `else` of the if block that checks to see if the value was specified in the input file.